feat(stages): checkpoint hashing stages into a new table

shekhirin opened the pull request.

mattsse reviewed (approved):
supportive of this,

pending @rkrasiuk

rkrasiuk commented on crates/stages/src/stages/hashing_account.rs:
```diff
@@ -168,27 +130,37 @@ impl<DB: Database> Stage<DB> for AccountHashingStage {
+            input.checkpoint.and_then(|checkpoint| checkpoint.account_hashing_stage_checkpoint());
+
+        if stage_checkpoint.is_some() {
+            debug!(target: "sync::stages::account_hashing::exec", checkpoint = ?stage_checkpoint, "Continuing inner account hashing checkpoint");
```
this debug log is valid only if the checkpoint range didn't change. pls move it below where we check that logic

rkrasiuk commented on crates/stages/src/stages/hashing_storage.rs:
```diff
@@ -89,27 +51,40 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
+            input.checkpoint.and_then(|checkpoint| checkpoint.storage_hashing_stage_checkpoint());
+
+        if stage_checkpoint.is_some() {
+            debug!(target: "sync::stages::storage_hashing::exec", checkpoint = ?stage_checkpoint, "Continuing inner storage hashing checkpoint");
```
same as the previous comment

rkrasiuk reviewed (commented):
did you try syncing this on sepolia/mainnet? ideal check would be to kill the app while it's in one of the hashing stages and restart to make sure it restores correctly

rkrasiuk reviewed (commented):
did you try syncing this on sepolia/mainnet? ideal check would be to kill the app while it's in one of the hashing stages and restart to make sure it restores correctly

shekhirin commented:
@rkrasiuk yep. Added 100ms sleep for hashing stages for ease of stop, synced to 100k mainnet blocks, dropped hashing stages, started node, stopped node while it executed account or storage hashing stage, then restarted.

rkrasiuk commented on crates/stages/src/stages/hashing_account.rs:
```diff
@@ -173,22 +135,33 @@ impl<DB: Database> Stage<DB> for AccountHashingStage {
+                    // and therefore should be hashed again. 
+                    if from == from_block && to == to_block =>
+                {
+                    if stage_checkpoint.is_some() {
```
this is redundant now

shekhirin replied:
oh right thanks, removed

rkrasiuk commented on crates/stages/src/stages/hashing_storage.rs:
```diff
@@ -94,22 +56,36 @@ impl<DB: Database> Stage<DB> for StorageHashingStage {
+                // and therefore should be hashed again. 
+                if from == from_block && to == to_block =>
+                    {
+                        if stage_checkpoint.is_some() {
```
same here

rkrasiuk reviewed (approved):
lgtm!
